mitiq/shadows/shadows.py

```diff
@@ -0,0 +1,198 @@
+
+    Returns:
+        A dictionary containing the bit strings, the Pauli strings
+        `bit_strings`: Circuit qubits computational basis
+        e.g. :math:`"01..":=|0\rangle|1\rangle..`.
+        `pauli_strings`: The local Pauli measurement performed on each
+        qubit. e.g."XY.." means perform local X-basis measurement on the
+        1st qubit, local Y-basis measurement the 2ed qubit in the circuit.
+    """
+    if random_seed is not None:
+        np.random.seed(random_seed)
+    r"""
+    Shadow stage 1: Sample random unitary form
+    :math:`\mathcal{g}\subset U(2^n)` and perform computational
+    basis measurement
+    """
+    # random Pauli measurement on the circuit
+    output = random_pauli_measurement(
+        circuit,
+        n_total_measurements=num_total_measurements_shadow,
+        executor=executor,
+        qubits=qubits,
+    )
+    return output
+
+
+def classical_post_processing(
+    shadow_outcomes: Tuple[List[str], List[str]],
+    rshadows: bool = False,
+    calibration_results: Optional[Dict[str, float]] = None,
+    observables: Optional[List[mitiq.PauliString]] = None,
+    k_shadows: Optional[int] = None,
+    state_reconstruction: Optional[bool] = False,
+) -> Mapping[str, Union[float, NDArray[Any]]]:
+    r"""
+    Executes a circuit with classical shadows. This function can be used for
+    state reconstruction or expectation value estimation of observables.
+
+    Args:
+        shadow_outcomes: The output of function `shadow_quantum_processing`.
+        rshadows: Whether to use the calibration results.
+        calibration_results: The output of function `pauli_twirling_calibrate`.
+        observables: The set of observables to measure.
+        k_shadows: Number of groups of "median of means" used for shadow
+            estimation of expectation values.
+        state_reconstruction: Whether to reconstruct the state or estimate
+            the expectation value of the observables.
+
+    Returns:
+        If state_reconstruction is True: state tomography matrix in
+        :math:`\mathbb{M}(\mathbb{C})_{2^n}` if rshadows is False,
+        otherwise state tomography vector in :math:`\mathbb{C}^{4^d}`.
+        If observables is given: estimated expectation values of
+        observables.
+    """
+
+    if rshadows:
+        if calibration_results is None:
+            raise ValueError(
+                "Calibration results cannot be None when rshadows"
+            )
+
+    """
+    Shadow stage 2: Estimate the expectation value of the observables OR
+    reconstruct the state
+    """
+    output: Dict[str, Union[float, NDArray[Any]]] = {}
+    if state_reconstruction:
+        reconstructed_state = shadow_state_reconstruction(
+            shadow_outcomes, rshadows, f_est=calibration_results
+        )
+        output["reconstructed_state"] = reconstructed_state  # type: ignore
+    elif (
+        observables is not None
+    ):  # Estimation expectation value of observables
```

Comment:
I think the comment can go on a newline to avoid breaking this statement over multiple lines (if you think it's necessary). I'm not sure it's helping me understand the code here, so if there's something you're trying to hint at, would you mind explaining a little more?

Reply:
I'll remove it. Thanks